README.rst
cmsplugin-blog Language Publish
===============================

An extension for `cmsplugin-blog <https://github.com/fivethreeo/cmsplugin-blog/>`_
which allows to publish only certain languages of a blog entry.


Installation
------------

You need to install the following prerequisites in order to use this app::

    pip install Django
    pip install django-cms
    pip install cmsplugin-blog
    pip install simple-translation

If you want to install the latest stable release from PyPi::

    $ pip install cmsplugin-blog-language-publish  (not available by now)

If you feel adventurous and want to install the latest commit from GitHub::

    $ pip install -e git://github.com/bitmazk/cmsplugin-blog-language-publish.git#egg=cmsplugin_blog_language_publish

Add ``cmsplugin_blog_language_publish`` to your ``INSTALLED_APPS``::

    INSTALLED_APPS = (
        ...,
        'cmsplugin_blog_language_publish',
    )


TODO: Add additional installation steps.


Usage
-----

TODO: Describe usage


Contribute
----------

If you want to contribute to this project, please perform the following steps::

    # Fork this repository
    # Clone your fork
    $ mkvirtualenv -p python2.7 cmsplugin-blog-language-publish
    $ pip install -r requirements.txt
    $ ./logger/tests/runtests.sh
    # You should get no failing tests

    $ git co -b feature_branch master
    # Implement your feature and tests
    # Describe your change in the CHANGELOG.txt
    $ git add . && git commit
    $ git push origin feature_branch
    # Send us a pull request for your feature branch

Whenever you run the tests a coverage output will be generated in
``tests/coverage/index.html``. When adding new features, please make sure that
you keep the coverage at 100%.


Roadmap
-------

Check the issue tracker on github for milestones and features to come.


CHANGELOG.txt
=== ongoing ===

- displaying the published state of each translation of Entry objects

=== 0.1.2 ===

- fixed class inheritance of CustomEntryAdmin

=== 0.1.1 ===

- fixed CustomEntryAdmin

=== 0.1 ===

Initial commit


cmsplugin_blog_language_publish/tests/factories.py
"""Factories for the models of the ``cmsplugin_blog_language_publish`` app."""
import factory

from cmsplugin_blog.models import Entry

from ..models import EntryLanguagePublish


class EntryFactory(factory.Factory):
    """Factory for the ``Entry`` model from ``cmsplugin_blog``."""
    FACTORY_FOR = Entry

    is_published = True


class EntryLanguagePublishFactory(factory.Factory):
    """Factory for the ``EntryLanguagePublish`` model."""
    FACTORY_FOR = EntryLanguagePublish

    entry = factory.SubFactory(EntryFactory)
    language = 'en'


cmsplugin_blog_language_publish/tests/models_tests.py
"""Tests for the models of the ``cmsplugin_blog_language_publish`` app."""
from django.test import TestCase

from cmsplugin_blog.models import Entry

from ..models import EntryLanguagePublish
from .factories import EntryLanguagePublishFactory


class EntryLanguagePublishTestCase(TestCase):
    """Tests for the ``EntryLanguagePublish`` model class."""
    longMessage = True

    def test_instantiation(self):
        """
        Test instantiation and save of the ``EntryLanguagePublish`` model.

        """
        EntryLanguagePublishFactory()
        self.assertEqual(EntryLanguagePublish.objects.all().count(), 1, msg=(
            'There should be one EntryLanguagePublish object.'))
        self.assertEqual(Entry.objects.all().count(), 1, msg=(
            'There should be one Entry object.'))


cmsplugin_blog_language_publish/tests/cmsplugin_blog_language_publish_tags_tests.py
"""Tests for the ``cmsplugin_blog_language_publish`` app's template tags."""
from django.test import TestCase

from ..templatetags.cmsplugin_blog_language_publish_tags import (
    get_published_entry)
from .factories import EntryLanguagePublishFactory


class GetPublishedEntryTestCase(TestCase):
    """Test for the ``get_published_entry`` template tag."""
    longMessage = True

    def setUp(self):
        self.entry_langpub = EntryLanguagePublishFactory(published=True)
        self.entry_not_published = EntryLanguagePublishFactory()

    def test_template_tag(self):
        """Test for the ``get_published_entry`` template tag."""
        entry = get_published_entry(self.entry_langpub.entry, 'en')
        self.assertEqual(entry, self.entry_langpub.entry, msg=(
            'When the correct language is published, the tag should return the'
            ' entry.'))

        entry = get_published_entry(self.entry_not_published.entry, 'en')
        self.assertIsNone(entry, msg=(
            'When the entry is not published in this language, it should'
            ' return None.'))


cmsplugin_blog_language_publish/admin.py
"""Admin classes for the ``cmsplugin_blog_language_publish`` app."""
from django.contrib import admin
from django.core.exceptions import ObjectDoesNotExist

from cmsplugin_blog.models import Entry
from cmsplugin_blog.admin import EntryAdmin, EntryForm
from simple_translation.translation_pool import translation_pool

from .models import EntryLanguagePublish


class EntryLanguagePublishAdmin(admin.ModelAdmin):
    """
    Admin for the ``EntryLanguagePublish`` model.

    Mainly used for debugging purposes. You would rather set the publish
    flag when editing a blog entry via the EntryAdmin.

    """
    list_display = (
        'title', 'language', 'is_published')

    def title(self, obj):
        return obj.entry_title.title

    def language(self, obj):
        return obj.entry_title.language


class CustomEntryForm(EntryForm):
    """
    Custom form for the ``Entry`` model, used by the ``EntryAdmin``.

    This is quite a hack. We will always set the `is_published` of the Entry
    model to true and we will override the initial value for the `is_published`
    field with the corresponding value of a ``EntryLanguagePublish`` object.

    """
    def __init__(self, *args, **kwargs):
        super(CustomEntryForm, self).__init__(*args, **kwargs)

        model = self._meta.model
        info = translation_pool.get_info(model)
        self.current_language = self.base_fields[info.language_field].initial
        title = None
        if self.instance and self.instance.pk:
            try:
                title = self.instance.entrytitle_set.get(
                    language=self.current_language)
            except ObjectDoesNotExist:
                pass

        publish = None
        if title:
            try:
                publish = EntryLanguagePublish.objects.get(
                    entry_title=title)
            except EntryLanguagePublish.DoesNotExist:
                pass

        self.initial['is_published'] = (
            publish and publish.is_published or False)

    def save(self, *args, **kwargs):
        """
        Setting ``is_published`` of the ``Entry`` always to ``True``.

        Since we no longer use this field in the admin, we will just always
        set it to true.

        """
        self.instance.is_published = True
        return super(CustomEntryForm, self).save(*args, **kwargs)


class CustomEntryAdmin(EntryAdmin):
    """
    Custom admin for the ``Entry`` model.

    We cannot display the ``is_published`` field in the list view any more
    because we would need to show one field per language now and that could get
    quite messy when there are many languages.

    """
    form = CustomEntryForm
    list_display = ('title', 'languages', 'author', 'pub_date')
    list_editable = ()
    list_filter = ('pub_date', )

    def save_model(self, request, obj, form, change):
        """
        Creating the ``EntryLanguagePublish`` model after saving the ``Entry``.

        Unfortunately this can't be done in the form's ``save()`` method. The
        admin only creates the ``EntryTitle`` object in this method. Therefore
        we override it and create our ``EntryLanguagePublish`` model here.

        """
        super(CustomEntryAdmin, self).save_model(request, obj, form, change)
        title = obj.entrytitle_set.get(language=form.current_language)
        published, created = EntryLanguagePublish.objects.get_or_create(
            entry_title=title)
        published.is_published = form.cleaned_data.get('is_published')
        published.save()


admin.site.unregister(Entry)
admin.site.register(Entry, CustomEntryAdmin)
admin.site.register(EntryLanguagePublish, EntryLanguagePublishAdmin)


cmsplugin_blog_language_publish/models.py
"""Models for the ``cmsplugin_blog_language_publish`` app."""
from django.conf import settings
from django.db import models
from django.utils.translation import ugettext_lazy as _


class EntryLanguagePublish(models.Model):
    """
    Holds the information about the published state of a certain language of
    an entry.

    :entry: FK to the actual entry.
    :language: The language that this state should be active for.
    :published: The published state of the entry and this language.

    """

    entry = models.ForeignKey(
        'cmsplugin_blog.Entry',
        verbose_name=_('Entry'),
        related_name='published_languages',
    )

    language = models.CharField(
        verbose_name=_('Language'),
        max_length=32,
        choices=settings.LANGUAGES,
    )

    published = models.BooleanField(
        verbose_name=_('Published'),
        default=False,
    )

    class Meta:
        unique_together = ('entry', 'language')


cmsplugin_blog_language_publish/migrations/0001_initial.py
# flake8: noqa
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Adding model 'EntryLanguagePublish'
        db.create_table('cmsplugin_blog_language_publish_entrylanguagepublish', (
            ('id', self.gf('django.db.models.fields.AutoField')(primary_key=True)),
            ('entry', self.gf('django.db.models.fields.related.ForeignKey')(related_name='published_languages', to=orm['cmsplugin_blog.Entry'])),
            ('language', self.gf('django.db.models.fields.CharField')(max_length=32)),
            ('pubished', self.gf('django.db.models.fields.BooleanField')(default=False)),
        ))
        db.send_create_signal('cmsplugin_blog_language_publish', ['EntryLanguagePublish'])

        # Adding unique constraint on 'EntryLanguagePublish', fields ['entry', 'language']
        db.create_unique('cmsplugin_blog_language_publish_entrylanguagepublish', ['entry_id', 'language'])


    def backwards(self, orm):
        # Removing unique constraint on 'EntryLanguagePublish', fields ['entry', 'language']
        db.delete_unique('cmsplugin_blog_language_publish_entrylanguagepublish', ['entry_id', 'language'])

        # Deleting model 'EntryLanguagePublish'
        db.delete_table('cmsplugin_blog_language_publish_entrylanguagepublish')


    models = {
        'cms.placeholder': {
            'Meta': {'object_name': 'Placeholder'},
            'default_width': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'slot': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'})
        },
        'cmsplugin_blog.entry': {
            'Meta': {'ordering': "('-pub_date',)", 'object_name': 'Entry'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_published': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'placeholders': ('djangocms_utils.fields.M2MPlaceholderField', [], {'to': "orm['cms.Placeholder']", 'symmetrical': 'False'}),
            'pub_date': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'tags': ('tagging.fields.TagField', [], {})
        },
        'cmsplugin_blog_language_publish.entrylanguagepublish': {
            'Meta': {'unique_together': "(('entry', 'language'),)", 'object_name': 'EntryLanguagePublish'},
            'entry': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'published_languages'", 'to': "orm['cmsplugin_blog.Entry']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'language': ('django.db.models.fields.CharField', [], {'max_length': '32'}),
            'pubished': ('django.db.models.fields.BooleanField', [], {'default': 'False'})
        }
    }

    complete_apps = ['cmsplugin_blog_language_publish']


cmsplugin_blog_language_publish/migrations/0002_auto__del_field_entrylanguagepublish_pubished__add_field_entrylanguage.py
# flake8: noqa
# -*- coding: utf-8 -*-
import datetime
from south.db import db
from south.v2 import SchemaMigration
from django.db import models


class Migration(SchemaMigration):

    def forwards(self, orm):
        # Deleting field 'EntryLanguagePublish.pubished'
        db.delete_column('cmsplugin_blog_language_publish_entrylanguagepublish', 'pubished')

        # Adding field 'EntryLanguagePublish.published'
        db.add_column('cmsplugin_blog_language_publish_entrylanguagepublish', 'published',
                      self.gf('django.db.models.fields.BooleanField')(default=False),
                      keep_default=False)


    def backwards(self, orm):
        # Adding field 'EntryLanguagePublish.pubished'
        db.add_column('cmsplugin_blog_language_publish_entrylanguagepublish', 'pubished',
                      self.gf('django.db.models.fields.BooleanField')(default=False),
                      keep_default=False)

        # Deleting field 'EntryLanguagePublish.published'
        db.delete_column('cmsplugin_blog_language_publish_entrylanguagepublish', 'published')


    models = {
        'cms.placeholder': {
            'Meta': {'object_name': 'Placeholder'},
            'default_width': ('django.db.models.fields.PositiveSmallIntegerField', [], {'null': 'True'}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'slot': ('django.db.models.fields.CharField', [], {'max_length': '50', 'db_index': 'True'})
        },
        'cmsplugin_blog.entry': {
            'Meta': {'ordering': "('-pub_date',)", 'object_name': 'Entry'},
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'is_published': ('django.db.models.fields.BooleanField', [], {'default': 'False'}),
            'placeholders': ('djangocms_utils.fields.M2MPlaceholderField', [], {'to': "orm['cms.Placeholder']", 'symmetrical': 'False'}),
            'pub_date': ('django.db.models.fields.DateTimeField', [], {'default': 'datetime.datetime.now'}),
            'tags': ('tagging.fields.TagField', [], {})
        },
        'cmsplugin_blog_language_publish.entrylanguagepublish': {
            'Meta': {'unique_together': "(('entry', 'language'),)", 'object_name': 'EntryLanguagePublish'},
            'entry': ('django.db.models.fields.related.ForeignKey', [], {'related_name': "'published_languages'", 'to': "orm['cmsplugin_blog.Entry']"}),
            'id': ('django.db.models.fields.AutoField', [], {'primary_key': 'True'}),
            'language': ('django.db.models.fields.CharField', [], {'max_length': '32'}),
            'published': ('django.db.models.fields.BooleanField', [], {'default': 'False'})
        }
    }

    complete_apps = ['cmsplugin_blog_language_publish']
